### ISSUE TITLE
Test npm package with Webpack-based project

### ISSUE THREAD

**Issue by Beak-man** [NASAWorldWind/WebWorldWind#157]
Testing Web Worldwind as an npm package within  [React Starter Kit](https://github.com/facebookincubator/create-react-app) is producing errors that point out to Web Worldwind's 3rd party libraries. This boilerplate project utilizes Webpack.

Some 3rd party libraries inside src/util/ were probably adapted from Node modules. These were modified as AMD modules in order to operate with the rest of the codebase, which is AMD-based. It also seems that their Node-related CommonJS interfaces were deliberately blocked, as it can be seen [here](https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/util/jszip.js#L12) and [here](https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/util/proj4-src.js#L1). [Edit: After unminifying, it can be seen that this is not the case. They're UMD interfaces. AMD and CommonJS are both supported].

Before projects like Browserify or Webpack that provide web browsers with (formerly backend only) CommonJS interfaces became prevalent, and in the middle of the explosion of different module definitions and package managers in the JavaScript world, this was a clever and quick workaround to make these libraries work in an AMD environment with the rest of the project without rewriting their interfaces.

According to testing however, this seems to be at odds with WebWorldWind's usage as an npm package destined to be used in the frontend (Issue #136), since these libraries fail to load while working in a Node environment in the browser (e.g. under Webpack).

Ways to provide these libraries with more up-to-date interfaces will be investigated.

It bears to say that this is also related to issue #147. Instead of going back and forth between CommonJS and AMD module interfaces, the ideal solution would be to substitute these libraries with npm dependencies in the project. This will necessitate architectural refactoring, however, and it will be considered in the future.

**Comment by Beak-man** [comment]
Some guidance:

https://github.com/webpack/webpack/issues/319

So maybe [these](https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/util/jszip.js#L12) [lines](https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/util/proj4-src.js#L1) are Browserify boilerplate?

The following project addressed something similar:

https://github.com/Matt-Esch/virtual-dom/pull/318/files?diff=split

**Comment by Beak-man** [comment]
Files proj4-src.js and jszip.js will be provided with AMD interfaces so to get them on line with the rest of the project's modules.

**Comment by Beak-man** [comment]
proj4-src is producing this error while trying to add the test npm package in the React Starter Kit:

> Error: formats/geojson/GeoJSONCRS missing util/proj4-src
    at main (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:324:27)
    at callDep (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:194:18)
    at main (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:319:31)
    at callDep (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:194:18)
    at main (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:319:31)
    at callDep (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:194:18)
    at req (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:359:20)
    at /Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:82515:12
    at defined (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:4:26)
    at Object.<anonymous> (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/test-web-world-wind/worldwind.js:12:2)
    at Module._compile (module.js:571:32)
    at Module._extensions..js (module.js:580:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/beakman/test_npm_project/ReactWorldwind/node_modules/babel-register/lib/node.js:152:7)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.25 (/Users/beakman/test_npm_project/ReactWorldwind/external "test-web-world-wind":1:1)
    at __webpack_require__ (/Users/beakman/test_npm_project/ReactWorldwind/webpack/bootstrap f9ee1423a858c7294e57:627:1)
    at fn (/Users/beakman/test_npm_project/ReactWorldwind/webpack/bootstrap f9ee1423a858c7294e57:47:1)
    at Object../src/components/Layout/Layout.js (/Users/beakman/test_npm_project/ReactWorldwind/build/chunks/home.js:696:78)

This is still happening after removing CommonJS interfaces. It's not module definition issue.

After further researching, it seems that Webpack has a problem with this library:

https://github.com/proj4js/proj4js/issues/214

**Comment by pdavidc** [comment]
Would you please share how you removed the CommonJS interfaces? It's best to get another pair of eyes on problems, rather than relying on yourself to ensure that everything has been done correctly.

**Comment by Beak-man** [comment]
Sure, at first I just gave precedence to AMD over CommonJS in the UMD interface.

This was the original (unminified):


    if ("object" == typeof exports) module.exports = e();
    else if ("function" == typeof define && define.amd) define(e);
    else {
        var f;
        "undefined" != typeof window ? f = window : "undefined" != typeof global ? f = global : "undefined" != typeof self && (f = self), f.proj4 = e()
    }

At first I just gave precedence to the AMD interface:

    if ("function" == typeof define && define.amd) define(e);
    else if ("object" == typeof exports) module.exports = e();
    else {
        var f;
        "undefined" != typeof window ? f = window : "undefined" != typeof global ? f = global : "undefined" != typeof self && (f = self), f.proj4 = e()
    }


Then I just removed the second line just to be sure that I'm never entering the CommonJS interface block. The library is exposing its AMD interface correctly.

The proj4js people mention that maybe is Webpack's configuration, but that this issue was corrected here: https://github.com/proj4js/proj4js/pull/224. I was about to try updating the library.

**Comment by pdavidc** [comment]
Thanks, Miguel. Attempting an update sounds good to me.

**Comment by Beak-man** [comment]
I updated it. They changed how the library interfaces:

```javascript
function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.proj4 = factory());
}
```

But the same thing is happening. I also tried commenting out the CommonJS interface just in case, to no avail.

**Comment by Beak-man** [comment]
The issue with proj4 was solved by removing the UMD pattern altogether and wrapping the library with an AMD interface exclusively, as in:

````javascript
define([], function () {
    // proj4 code
});
````

Which I think was what you had in mind in the beginning @pdavidc. It seems that this is a common problem with proj4's UMD interface. I found, for instance, [a guy who did almost the same thing](https://github.com/temsa/node-proj4js), but in this case to only leave a CommonJS interface.

I hit another glitch with Webpack after this. The images in the local folder aren't being loaded, so we have no compass, BMNGOneImageLayer, etc. (this can be seen in [the React tester I did](https://github.com/Beak-man/worldwind-react)). I'm looking into seeing if, again, Webpack is opinionated on how to retrieve local files, [as it can be seen here.](https://stackoverflow.com/questions/34582405/react-wont-load-local-images)

**Comment by pdavidc** [comment]
@Beak-man This is indeed one of the possible solutions I had in mind for proj4, and any other libraries in src/util that exhibit issues when loaded via NPM.

It sounds like you've solved the AMD issue only to find another relating to resource loading. Do you think the two issues are independent? If so, let's wrap up the AMD issue one by making the necessary changes to the libraries included in src/util, test those changes, then submit a pull request.

**Comment by Beak-man** [comment]
@pdavidc I think the resource loading problem is related to interfacing, but independent to the AMD issue with proj4. Curiously, it _only_ happens with proj4. Jszip, has no issues whatsoever, and now our package can be loaded inside Node, even (it runs into errors when WorldWind tries to access a DOM object, of course). That being said, as soon as I'm able to load local resources, I will check that jszip is working properly.

I will do as you suggested and try to build a simple npm package that retrieves a file locally.

**Comment by Beak-man** [comment]
Doing a tiny Webpack-based project and loading Web WorldWind's pre-release npm package points towards a relative path issue in regards to the local image files. The image files are not being retrieved from the /images folder located in the root folder of the package when it's installed as a node_module, but from the root of the web document itself. For some reason this is not causing a 404 issue in the React tester, but it does in the tiny project where I tested it. This will be addressed in issue #164. Tests with Webpack can be considered done.
